StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/FixAllContextHelper.cs

```diff
@@ -0,0 +1,307 @@
+        }
+
+        public static async Task<ImmutableDictionary<Project, ImmutableArray<Diagnostic>>> GetProjectDiagnosticsToFixAsync(FixAllContext fixAllContext)
+        {
+            var project = fixAllContext.Project;
+            if (project != null)
+            {
+                switch (fixAllContext.Scope)
+                {
+                case FixAllScope.Project:
+                    var diagnostics = await fixAllContext.GetProjectDiagnosticsAsync(project).ConfigureAwait(false);
+                    return ImmutableDictionary<Project, ImmutableArray<Diagnostic>>.Empty.SetItem(project, diagnostics);
+
+                case FixAllScope.Solution:
+                    var projectsAndDiagnostics = new ConcurrentDictionary<Project, ImmutableArray<Diagnostic>>();
+                    var options = new ParallelOptions() { CancellationToken = fixAllContext.CancellationToken };
+                    Parallel.ForEach(project.Solution.Projects, options, proj =>
+                    {
+                        fixAllContext.CancellationToken.ThrowIfCancellationRequested();
+                        var projectDiagnosticsTask = fixAllContext.GetProjectDiagnosticsAsync(proj);
+                        projectDiagnosticsTask.Wait(fixAllContext.CancellationToken);
+                        var projectDiagnostics = projectDiagnosticsTask.Result;
+                        if (projectDiagnostics.Any())
+                        {
+                            projectsAndDiagnostics.TryAdd(proj, projectDiagnostics);
+                        }
+                    });
+
+                    return projectsAndDiagnostics.ToImmutableDictionary();
+                }
+            }
+
+            return ImmutableDictionary<Project, ImmutableArray<Diagnostic>>.Empty;
+        }
+
+        public static async Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync(Compilation compilation, CompilationWithAnalyzers compilationWithAnalyzers, ImmutableArray<DiagnosticAnalyzer> analyzers, IEnumerable<Document> documents, bool includeCompilerDiagnostics, CancellationToken cancellationToken)
+        {
+            if (GetAnalyzerSyntaxDiagnosticsAsync == null || GetAnalyzerSemanticDiagnosticsAsync == null)
+            {
+                // In everything except Roslyn 1.1, we use GetAllDiagnosticsAsync and return the result.
+                var allDiagnostics = await compilationWithAnalyzers.GetAllDiagnosticsAsync().ConfigureAwait(false);
+                return allDiagnostics;
+            }
+
+            compilationWithAnalyzers.Compilation.GetDeclarationDiagnostics(cancellationToken);
+
+            // Note that the following loop to obtain syntax and semantic diagnostics for each document cannot operate
+            // on parallel due to our use of a single CompilationWithAnalyzers instance.
+            var diagnostics = ImmutableArray<Diagnostic>.Empty;
```

Comment:
:thought_balloon: using an `ImmutableArray<Diagnostic>.Builder` or a `List<Diagnostic>` would improve the performance when adding diagnostics. There is no need to have the diagnostics in the `ImmutableArray<Diagnostic>` type yet, as they are never passed to another method.